File: LeetCode/19. Remove Nth Node From End of List/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        ListNode slow = dummy;
        for (int i = 0; i < n; i++) {//create gap between fast and slow
            fast = fast.next;
        }
        while (fast.next != null) {//to the end, then delete the one after slow
            fast = fast.next;
            slow = slow.next;
        }
        slow.next = slow.next.next;
        return dummy.next;
    }
}

File: LeetCode/19. Remove Nth Node From End of List/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        left = right = head
        for i in range(n):
            right = right.next
        if not right:
            return left.next
        while right.next:
            left, right = left.next, right.next
        left.next = left.next.next
        return head